src/main/java/co/cask/gcp/spanner/sink/SpannerSink.java

```diff
@@ -79,10 +114,79 @@ public void prepareRun(BatchSinkContext context) {
-                                    fields.stream().map(Schema.Field::getName).collect(Collectors.toList()));
+      // Record the field level WriteOperation
+      lineageRecorder.recordWrite("Write", "Wrote to Spanner table.",
+                                  fields.stream().map(Schema.Field::getName).collect(Collectors.toList()));
+    }
+  }
+
+  private void createTableIfNotPresent(DatabaseClient dbClient, Database database) {
+    boolean tableExists = isTablePresent(dbClient);
+    if (!tableExists) {
+      String createStmt = SpannerUtil.convertSchemaToCreateStatement(config.table, config.getSchema());
+      LOG.debug("Creating table with create statement: {} in database {} of instance {}", createStmt,
+                config.database, config.instance);
+      // In Spanner table creation is an update ddl operation on the database.
+      Operation<Void, UpdateDatabaseDdlMetadata> op = database.updateDdl(Collections.singletonList(createStmt), null);
+      // Table creation is an async operation. So wait until table is created.
+      op.waitFor().getResult();
+    }
+  }
+
+  private boolean isTablePresent(DatabaseClient dbClient) {
+    // Spanner does not have apis to get table or check if a given table exists. So select the table name from
+    // information schema (metadata) of spanner database.
+    ResultSet resultSet = dbClient
+      .singleUse()
+      .executeQuery(Statement.of("SELECT\n" +
+                                   "    t.table_name\n" +
+                                   "FROM\n" +
+                                   "    information_schema.tables AS t\n" +
+                                   "WHERE\n" +
+                                   "    t.table_catalog = '' AND t.table_schema = '' AND\n" +
+                                   "    t.table_name = '" + config.table + "'"));
+    boolean tableExists = false;
+    while (resultSet.next()) {
+      tableExists = true;
+      break;
+    }
+    // close result set to free up resources.
+    resultSet.close();
+    return tableExists;
+  }
+
+  private Database getOrCreateDatabase(DatabaseAdminClient dbAdminClient) {
+    Database database = getDatabaseIfPresent(dbAdminClient);
+
+    if (database == null) {
+      LOG.debug("Database not found. Creating database {} in instance {}.", config.database, config.instance);
+      // Create database
+      Operation<Database, CreateDatabaseMetadata> op =
+        dbAdminClient.createDatabase(config.instance, config.database, Collections.emptyList());
+      // database creation is an async operation. Wait until database creation operation is complete.
+      database = op.waitFor().getResult();
+    }
+
+    return database;
+  }
+
+  @Nullable
+  private Database getDatabaseIfPresent(DatabaseAdminClient dbAdminClient) {
+    Database database = null;
+    Page<Database> page = dbAdminClient.listDatabases(config.instance, Options.pageSize(100));
```

Comment:
there's a getDatabase() method that should be used instead.

Reply:
It throws SpannerException: `Database does not exist`. We can catch that exception and then create a table. I thought this way would be cleaner. Please let me know if you think otherwise.

Reply:
Can you determine that it was a not found exception through the error code? I would hope that information is in the exception and you don't have to parse a message.

Reply:
yea I think they do have error code for not_found. Will make that change.